### PR TITLE
feat: Quantile function in SQL

### PR DESCRIPTION
**Pull request by pomo-mondreganto**
This implements a feature requested in https://github.com/pola-rs/polars/issues/7227: quantile functions in SQL. I also went ahead and added median function tests as those were missing.

### CODE REVIEW

**Review by alexander-beedie — CHANGES_REQUESTED** [review]
FYI: there is no `QUANTILE` function in PostgreSQL (or ANSI SQL), which is what we usually aim to match 😉 Can you reformulate this to expose one (or both) of the standard PostgreSQL functions, `PERCENTILE_DISC` and/or `PERCENTILE_CONT`[^1] instead? I believe you will also need to integrate parsing support for `WITHIN GROUP` to conform to the expected syntax.

This would be a great addition to our SQL interface and I'd be very happy to see this capability added 👍

Looking at DuckDB they expose `QUANTILE_DISC` and `QUANTILE_CONT`[^2] in addition to the two percentile functions above, so that is probably the way to go for quantile funcs. As long as we have the standard SQL funcs I'm not averse to adding those as well.

[^1]: https://www.postgresql.org/docs/current/functions-aggregate.html#FUNCTIONS-ORDEREDSET-TABLE
[^2]: https://duckdb.org/docs/sql/functions/aggregates#statistical-aggregates

**Comment by pomo-mondreganto** [comment]
I believe I need to gain a deeper understanding of the SQL engine used to implement something more complex than a function itself (required for `WITHIN GROUP` support), so I'll focus on QUANTILE_* functions in this PR. I also might open a new PR for PERCENTILE_* in the near future. 

Related question: I was wondering if it's worth to leave my proposed API as-is (or only leave the `QUANTILE(value, interpolation)`) to expose all the available interpolation algorithms polars has in addition to _disc (nearest?) and _cont (linear?) quantiles.

**Comment by alexander-beedie** [comment]
> I believe I need to gain a deeper understanding of the SQL engine used to implement something more complex than a function itself (required for `WITHIN GROUP` support), so I'll focus on QUANTILE_* functions in this PR. I also might open a new PR for PERCENTILE_* in the near future.

No problem! And I agree, adding support for the other syntax is a larger PR 😆 

> Related question: I was wondering if it's worth to leave my proposed API as-is (or only leave the `QUANTILE(value, interpolation)`) to expose all the available interpolation algorithms polars has in addition to _disc (nearest?) and _cont (linear?) quantiles.

No, because it has no analog in any standard SQL syntax; we aren't just using the SQL interface to expose custom Polars functions with a custom API, because otherwise that's not really SQL, it's a custom language. Since there _are_ other SQL examples of `QUANTILE_DISC` and `QUANTILE_CONT` (in both DuckDB and Databend) then I think it's reasonable to expose it this way 👌

**Comment by pomo-mondreganto** [comment]
Is it ok now? I have an internal feature in the company depending on this and I would like to at least establish the interface to avoid refactoring client's code later. I was also wondering what's the release policy of the rust crates and whether I should stay on git dependencies for the near future?

P.S. I've looked though sqlparser and it seems like WITH GROUP syntax is supported out of the box, so this next PR with percentiles might be coming sooner that I though.

**Comment by alexander-beedie** [comment]
> Is it ok now? I have an internal feature in the company depending on this and I would like to at least establish the interface to avoid refactoring client's code later. I was also wondering what's the release policy of the rust crates and whether I should stay on git dependencies for the near future?

The syntax looks good to me, and I just gave it a quick test to sanity-check things and it's returning the results I expect. Will give it a more thorough test (edge-cases, behaviour with nulls, etc) tomorrow.

One thing I spotted - at the moment `QUANTILE_DISC` returns float values even if the input is integer; as the results are guaranteed to be discrete values from the original data then it should preserve the original dtype.

If you can fix that (and further testing look good tomorrow) then this should be good merge shortly 👍
Example:
```python
df.sql(
  """
  SELECT  
    QUANTILE_DISC(Sales, 0.10) as q10,
    QUANTILE_DISC(Sales, 0.25) as q25,
    QUANTILE_DISC(Sales, 0.40) as q40,
    QUANTILE_DISC(Sales, 0.55) as q55,
    QUANTILE_DISC(Sales, 0.70) as q70,
    QUANTILE_DISC(Sales, 0.85) as q85
  FROM self 
  """
)
# ┌────────┬────────┬────────┬────────┬────────┬────────┐
# │ q10    ┆ q25    ┆ q40    ┆ q55    ┆ q70    ┆ q85    │
# │ ---    ┆ ---    ┆ ---    ┆ ---    ┆ ---    ┆ ---    │
# │ f64    ┆ f64    ┆ f64    ┆ f64    ┆ f64    ┆ f64    │  << should be int
# ╞════════╪════════╪════════╪════════╪════════╪════════╡
# │ 1000.0 ┆ 2000.0 ┆ 3000.0 ┆ 3000.0 ┆ 4000.0 ┆ 5000.0 │
# └────────┴────────┴────────┴────────┴────────┴────────┘
```

> P.S. I've looked through sqlparser and it seems like WITH GROUP syntax is supported out of the box, so this next PR with percentiles might be coming sooner that I though.

Yes, it's available in `sqlparser-rs`, we just don't have any integration on our side yet.

**Comment by pomo-mondreganto** [comment]
Do you suggest casting the result series to the source series datatype explicitly? That sure comes with some performance penalty, is it acceptable in this case?

**Comment by pomo-mondreganto** [comment]
I believe casting to the original series dtype is blocked by https://github.com/pola-rs/polars/issues/4982, I've found no way to check the original dtype as well.

**Comment by alexander-beedie** [comment]
> I believe casting to the original series dtype is blocked by #4982, I've found no way to check the original dtype as well.

Ok; let's not worry about that for now - we can merge as float and update if necessary.

I've found a few cases that don't seem to line up with DuckDB's implementation (for the DISC variant; CONT seems good) which need some explanation (and/or a fix?):

```python
from polars.testing import assert_frame_equal
import polars as pl
import duckdb as dd

df = pl.DataFrame({'n': [
    1.523029,
    0.767434,
    0.496714,
    0.647688,]
})
```
```python
sql_disc = f"SELECT QUANTILE_DISC(n, 0.76) AS n FROM df"
print(dd.sql(sql_disc))
# ┌──────────┐
# │    n     │
# │  double  │
# ├──────────┤
# │ 1.523029 │
# └──────────┘

print(pl.sql(sql_disc).collect())
# shape: (1, 1)
# ┌──────────┐
# │ n        │
# │ ---      │
# │ f64      │
# ╞══════════╡
# │ 0.767434 │
# └──────────┘
```

**Review by alexander-beedie — CHANGES_REQUESTED** [review]
> I believe casting to the original series dtype is blocked by #4982, I've found no way to check the original dtype as well.

Ok; let's not worry about that for now - we can merge as float and update if necessary.

I've found a few cases that don't seem to line up with DuckDB's implementation (for the DISC variant; CONT seems good) which need some explanation (and/or a fix):

For example:
```python
import polars as pl
import duckdb as dd

df = pl.DataFrame({'n': [
    1.523029,
    0.767434,
    0.647688,
    0.496714,
]})
```
```python
sql_disc = "SELECT QUANTILE_DISC(n, 0.60) AS q FROM df"
print(dd.sql(sql_disc))
# ┌──────────┐
# │    n     │
# │  double  │
# ├──────────┤
# │ 0.767434 │
# └──────────┘

print(pl.sql(sql_disc).collect())
# shape: (1, 1)
# ┌──────────┐
# │ n        │
# │ ---      │
# │ f64      │
# ╞══════════╡
# │ 0.647688 │
# └──────────┘
```
About half of the percentiles I tried with `QUANTILE_DISC` (from 0.01 to 0.99) returned a different result.

**Comment by pomo-mondreganto** [comment]
I believe DuckDB is using `nearest` quantile where's my impl uses `lower`. That shouldn't be a problem, I'll change it to 'nearest' and add some compliance test cases then.

**Comment by alexander-beedie** [comment]
> I believe DuckDB is using `nearest` quantile where's my impl uses `lower`. That shouldn't be a problem, I'll change it to 'nearest' and add some compliance test cases then.

Sounds good; I think it's worth returning the same thing here by default 👌

**Comment by pomo-mondreganto** [comment]
After a brief vacation, back to this PR. My conformance test is currently set up the following way: both systems have an integer column `Sales` consisting of 6 numbers `1000 * (i + 1)`:

```rust
df! {
  "Year" => [2018, 2018, 2019, 2019, 2020, 2020],
  "Country" => ["US", "UK", "US", "UK", "US", "UK"],
  "Sales" => [1000, 2000, 3000, 4000, 5000, 6000]
}
```

```
D select * from kek;
┌───────┐
│ sales │
│ int32 │
├───────┤
│  1000 │
│  2000 │
│  3000 │
│  4000 │
│  5000 │
│  6000 │
└───────┘
```

DuckDB thinks that 0.7-th discrete quantile of the table is `5000`:

```
D select quantile_disc(sales, 0.7) from kek;
┌───────────────────────────┐
│ quantile_disc(sales, 0.7) │
│           int32           │
├───────────────────────────┤
│                      5000 │
└───────────────────────────┘
```

Whereas my implementation results in `4000` with `lower` interpolation method. 

DuckDB [docs](https://duckdb.org/docs/sql/functions/aggregates.html#quantile_discx-pos) state that the function should return `the floor(pos * (n_nonnull_values - 1))th (zero-indexed) element`, so for our case it's `floor(0.7 * (6 - 1))`, which is 3, so it should return 3rd 0-indexed element, which is 4000 and also what my implementation does. I believe this to be a bug in DuckDB. 

My suggestion is to commit the "lower" interpolation method with my current tests and merge it to the upstream.

**Comment by pomo-mondreganto** [comment]
@alexander-beedie wdyt?

**Comment by alexander-beedie** [comment]
> @alexander-beedie wdyt?

Sounds good to me; have you reported the bug to the DuckDB folks? Doesn't stop us moving forward here, but would be interested in their take on it 👍

**Comment by pomo-mondreganto** [comment]
I've pushed my conformance tests (and also support for 0 and 1 quantiles as they're parsed as integers). I'll open an issue for DuckDB in a few days when I have time to do a more proper test (from DuckDB master branch) and mention this PR in that issue. As for now -- I suggest merging after the final review.

**Comment by pomo-mondreganto** [comment]
@alexander-beedie :(

**Comment by pomo-mondreganto** [comment]
I've also reported the issue to duckdb: https://github.com/duckdb/duckdb/issues/14144

**Comment by pomo-mondreganto** [comment]
@ritchie46 @orlp can anyone look at this in the meantime?

**Comment by soerenwolfers** [comment]
@pomo-mondreganto duckdb's documentation was wrong. I just fixed it, so you can look up the correct formula in a few minutes, but intuitively it's quite clear what `quantile_disc(x, q)` is trying to do: Split up the interval `[0, 1] ` in as many equal-length intervals as you have elements, then return the `i`th element if `q` is in the `i`th interval. In your case, the intervals are `[0,1/6], (1/6, 2/6], (2/6, 3/6],(3/6, 4/6],(4/6, 5/6], (5/6, 1]` so without going into detail how edge cases are resolved (left open right closed intervals) it's clear that `0.7` is in the penultimate interval and the answer therefore has to be the penultimate item, `5000`.

**Comment by pomo-mondreganto** [comment]
That sounds quite reasonable, thanks for the explanation. The formula you provided is indeed correct. However, the polars backend lacks the required "quantile interpolation" method in its execution backend [here](https://github.com/pola-rs/polars/blob/57df7b0df6cb53e8987f3246356c107aeb26fff9/crates/polars-core/src/chunked_array/ops/aggregate/quantile.rs#L21):

```rust
    let float_idx = ((length - null_count) as f64 - 1.0) * quantile + null_count as f64;
    let mut base_idx = match interpol {
        QuantileInterpolOptions::Nearest => {
            let idx = float_idx.round() as usize;
            return (float_idx.round() as usize, 0.0, idx);
        },
        QuantileInterpolOptions::Lower
        | QuantileInterpolOptions::Midpoint
        | QuantileInterpolOptions::Linear => float_idx as usize,
        QuantileInterpolOptions::Higher => float_idx.ceil() as usize,
    };

    base_idx = base_idx.clamp(0, length - 1);
    let top_idx = f64::ceil(float_idx) as usize;
``` 

It's easy to see that it'll not be compatible with duckdb's implementation in any case. The problem above was with `q=0.7` and `QuantileInterpolOptions::Linear`, and switching to `Linear` or `Higher` would break the `q=0.1` case as `(6 - 1) * 0.1 = 0.5`, both `round` and `ceil` would return 1, whereas duckdb returns 0.

I could change how this function works in polars (e.g. use the `((length - null_count) as f64) * quantile - 1.0` for `float_idx` instead, but that would mean a backward-incompatible change in `core` and I sure would need an approval from any code owner beforehand. Please respond, what do you think?

**Comment by alexander-beedie** [comment]
Apologies for the delay, I was travelling, then sick, and then massively busy at work 😓

So, just to clarify the current state:

* We would need a new option in `QuantileInterpolOptions` to match? Or to make a more substantial change? 
* If the former, that doesn't sound unreasonable, as we could expose the new option everywhere.
* The `QUANTILE_CONT` function matches the expected values, only `QUANTILE_DISC` needs a decision.

If you wanted to break the PR into two pieces (so we don't delay merging the part that works any longer) then we could approve/merge `QUANTILE_CONT`, and make a separate PR/Issue for `QUANTILE_DISC`? 👍

**Comment by pomo-mondreganto** [comment]
Yeah, let's go with a new option. I'll update this PR today to exclude `quantile_disc` to merge quantile_cont and return with quantile_disc when it's ready. Seems like I'll need to change a ton of code to handle the new option, will take a while.

**Comment by pomo-mondreganto** [comment]
Done, `QUANTILE_CONT` should be ready to merge now, @alexander-beedie

**Comment by soerenwolfers** [comment]
Frankly, polars should consider changing its discrete interpolation options. They seem to be borne out of the same misunderstanding that bore the previously incorrect duckdb documentation (which I had written myself, so no blame here): That discretization should obviously just mean snapping the continuous index to an integer index. In reality, however, none of the currently offered options achieve the one thing you'd most certainly expect from a discrete quantile: That each element is returned with equal chance, for a randomly drawn `q` in `[0, 1]`. In other words, they fail to achieve the main goal of quantiles: To partition the values of the given distribution into equally "large" subgroups. In fact, the `Lower` and `Higher` options essentially never return one of the values, whereas the `Nearest` option returns the boundary elements half as frequently as the other elements.

**Comment by pomo-mondreganto** [comment]
I've actually finished implementing quantile_disc using the new quantile interpolation method "bucket", but waiting for this PR to be merged to have less merge conflicts before opening a new one.

File: crates/polars-sql/src/functions.rs
Comment (alexander-beedie) [inline crates/polars-sql/src/functions.rs:697]: Think this should be `quantile_cont` 🤔

**Review by alexander-beedie — CHANGES_REQUESTED** [review]
One (very!) minor update needed, then this looks good to me 👍

**Review by alexander-beedie — APPROVED** [review]

**Comment by alexander-beedie** [comment]
All good now - many thanks for this! 

Based on this comment (https://github.com/pola-rs/polars/pull/18047#issuecomment-2387998726) it sounds like we should discuss the current quantile implementation before moving on to `QUANTILE_DISC` 🤔 (@soerenwolfers, thanks for the feedback).

**Pull request merged**
